src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNITcpHandle.cs

```diff
@@ -182,31 +183,72 @@ public SNITCPHandle(string serverName, int port, long timerExpire, object callba
             _status = TdsEnums.SNI_SUCCESS;
         }
 
-        private static async Task<Socket> ConnectAsync(string serverName, int port)
+        private static Socket Connect(string serverName, int port, TimeSpan timeout)
         {
-            IPAddress[] addresses = await Dns.GetHostAddressesAsync(serverName).ConfigureAwait(false);
-            IPAddress targetAddrV4 = Array.Find(addresses, addr => (addr.AddressFamily == AddressFamily.InterNetwork));
-            IPAddress targetAddrV6 = Array.Find(addresses, addr => (addr.AddressFamily == AddressFamily.InterNetworkV6));
-            if (targetAddrV4 != null && targetAddrV6 != null)
+            IPAddress[] ipAddresses = Dns.GetHostAddresses(serverName);
+            IPAddress serverIPv4 = null;
+            IPAddress serverIPv6 = null;
+            foreach (IPAddress ipAdress in ipAddresses)
             {
-                return await ParallelConnectAsync(new IPAddress[] { targetAddrV4, targetAddrV6 }, port).ConfigureAwait(false);
+                if (ipAdress.AddressFamily == AddressFamily.InterNetwork)
+                {
+                    serverIPv4 = ipAdress;
+                }
+                else if (ipAdress.AddressFamily == AddressFamily.InterNetworkV6)
+                {
+                    serverIPv6 = ipAdress;
+                }
             }
-            else
-            {
-                IPAddress targetAddr = (targetAddrV4 != null) ? targetAddrV4 : targetAddrV6;
-                var socket = new Socket(targetAddr.AddressFamily, SocketType.Stream, ProtocolType.Tcp);
+            ipAddresses = new IPAddress[] { serverIPv4, serverIPv6 };
+            Socket[] sockets = new Socket[2];
 
-                try
+            CancellationTokenSource cts = new CancellationTokenSource();
+            cts.CancelAfter(timeout);
+            void Cancel()
+            {
+                for (int i = 0; i < sockets.Length; ++i)
                 {
-                    await socket.ConnectAsync(targetAddr, port).ConfigureAwait(false);
+                    try
+                    {
+                        if (sockets[i] != null && !sockets[i].Connected)
+                        {
+                            sockets[i].Dispose();
```

Comment:
Do we need lock statements when disposing/nulling sockets[i]? Could get a NullRefException if the dispose blocks get interleaved.

Reply:
Although it looks like lock() throws an ArgNullException if the object being locked is null. Might have to store sockets[i] in a local variable first before locking on it.

Reply:
Try/Catch block was used for that reason. I also considered the locking or event based synchronization to prevent exception, but it affects to performance. We are all starve for performance on .NET Core in Linux, and locking or event approach does not seem to be the solution. Catching null pointer exception would not be so fancy, but it is barely lowers speed.

Reply:
Please, consider this order:
1) **if (sockets[i] != null && !sockets[i].Connected)** is true
2) Socket opens connection
3) **if (sockets[i] != null)** gives true
4) **if (sockets[i].Connected)** gives true
5) **sockets[i].Dispose();**
6) **availableSocket = sockets[i];**
7) **break; return availableSocket;**